File: maui-app/Services/DownloadService.cs
using System.Collections.Concurrent;

namespace MauiFirebaseMetrics.Services;

public class DownloadService
{
    private readonly CrashService _crashService;
    private readonly ConcurrentDictionary<string, DownloadMetrics> _activeDownloads = new();

    public DownloadService(CrashService crashService)
    {
        _crashService = crashService;
    }

    public void OnDownloadStarted(string downloadId, string url, long? expectedBytes)
    {
        _activeDownloads.TryAdd(downloadId, new DownloadMetrics
        {
            Url = url,
            ExpectedBytes = expectedBytes ?? -1,
            StartedAt = DateTime.UtcNow
        });

        _crashService.Log($"Download started: {url} (expected: {expectedBytes?.ToString() ?? "unknown"} bytes)");
        _crashService.SetMetadata("last_download_url", url);
    }

    public void OnDownloadProgress(string downloadId, long bytesReceived, long totalBytes)
    {
        if (!_activeDownloads.TryGetValue(downloadId, out var metrics))
            return;

        metrics.BytesReceived = bytesReceived;

        if (totalBytes > 0)
        {
            var percent = (int)((double)bytesReceived / totalBytes * 100);
            if (percent % 25 == 0)
            {
                _crashService.Log($"Download progress: {percent}% ({bytesReceived}/{totalBytes} bytes)");
            }
        }
    }

    public void OnDownloadCompleted(string downloadId, string localPath)
    {
        if (!_activeDownloads.TryRemove(downloadId, out var metrics))
            return;

        var durationMs = (DateTime.UtcNow - metrics.StartedAt).TotalMilliseconds;
        var throughputKbps = metrics.BytesReceived > 0 && durationMs > 0
            ? (metrics.BytesReceived / 1024.0) / (durationMs / 1000.0)
            : 0;

        _crashService.Log(
            $"Download complete: {metrics.Url} — " +
            $"{metrics.BytesReceived} bytes in {durationMs:F0}ms " +
            $"({throughputKbps:F1} KB/s)");

        _crashService.SetMetadata("last_download_duration_ms", ((long)durationMs).ToString());
        _crashService.SetMetadata("last_download_throughput_kbps", ((long)throughputKbps).ToString());
    }

    public void OnDownloadFailed(string downloadId, Exception error)
    {
        if (_activeDownloads.TryRemove(downloadId, out var metrics))
        {
            _crashService.SetMetadata("failed_download_url", metrics.Url);
        }

        _crashService.RecordNonFatal(error, context: "WKDownloadDelegate.DidFail");
    }

    private sealed class DownloadMetrics
    {
        public required string Url { get; set; }
        public long ExpectedBytes { get; set; }
        public long BytesReceived { get; set; }
        public DateTime StartedAt { get; set; }
    }
}

File: react-native-app/services/download-service.ts
import {crashService} from './crash-service';

interface DownloadMetrics {
  url: string;
  expectedBytes: number;
  bytesReceived: number;
  startedAt: number;
}

class DownloadService {
  private activeDownloads: Map<string, DownloadMetrics> = new Map();

  onDownloadStarted(downloadId: string, url: string, expectedBytes?: number): void {
    this.activeDownloads.set(downloadId, {
      url,
      expectedBytes: expectedBytes ?? -1,
      bytesReceived: 0,
      startedAt: Date.now(),
    });

    crashService.log(`Download started: ${url} (expected: ${expectedBytes ?? 'unknown'} bytes)`);
    crashService.setMetadata('last_download_url', url);
  }

  onDownloadProgress(downloadId: string, bytesReceived: number, totalBytes: number): void {
    const metrics = this.activeDownloads.get(downloadId);
    if (!metrics) return;

    metrics.bytesReceived = bytesReceived;

    if (totalBytes > 0) {
      const percent = Math.floor((bytesReceived / totalBytes) * 100);
      if (percent % 25 === 0) {
        crashService.log(`Download progress: ${percent}% (${bytesReceived}/${totalBytes} bytes)`);
      }
    }
  }

  onDownloadCompleted(downloadId: string, _localPath: string): void {
    const metrics = this.activeDownloads.get(downloadId);
    if (!metrics) return;
    this.activeDownloads.delete(downloadId);

    const durationMs = Date.now() - metrics.startedAt;
    const throughputKbps =
      metrics.bytesReceived > 0 && durationMs > 0
        ? (metrics.bytesReceived / 1024) / (durationMs / 1000)
        : 0;

    crashService.log(
      `Download complete: ${metrics.url} — ` +
        `${metrics.bytesReceived} bytes in ${durationMs}ms ` +
        `(${throughputKbps.toFixed(1)} KB/s)`,
    );

    crashService.setMetadata('last_download_duration_ms', String(Math.round(durationMs)));
    crashService.setMetadata('last_download_throughput_kbps', String(Math.round(throughputKbps)));
  }

  onDownloadFailed(downloadId: string, error: Error): void {
    const metrics = this.activeDownloads.get(downloadId);
    if (metrics) {
      this.activeDownloads.delete(downloadId);
      crashService.setMetadata('failed_download_url', metrics.url);
    }

    crashService.recordNonFatal(error, 'DownloadDelegate.DidFail');
  }
}

export const downloadService = new DownloadService();
